### PR TITLE
Added common module that defines weak methods that can be overriden b…

### PR DESCRIPTION
**Pull request by vijs**
#### Problem
* Example lighting-app example does not allow external/third-party vendors to override default Device Attestation Provider and Verifier.

#### Change Overview
* Added common module that defines weak methods that can be overridden by vendor apps - that return the default Attestation Trust Stores and Example DAC Provider.
* Updated Linux lighting-app example to generate a static library.
* Updated main Linux app platform code to use the newly added weak methods to fetch the Attestation Trust Store and DAC Provider.

#### Testing
* Matter Unit Tests
* Tested Commissioning flow (chip-tool vs lighting-app)

### CODE REVIEW

File: examples/common/datt-config/DattConfig.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <DattConfig.h>
+
+__attribute__((weak)) const chip::Credentials::AttestationTrustStore * GetAttestationTrustStore()
```
Comment (Damian-Nordic) [inline examples/common/datt-config/DattConfig.cpp:21]: Why is this needed? Can't an application just pass a different object to `SetDeviceAttestationCredentialsProvider` instead of `chip::Credentials::Examples::GetExampleDACProvider()`?

Comment (vijs) [reply examples/common/datt-config/DattConfig.cpp:21]: Currently, if a developer wants to override the default DAC Provider, he will need to edit the example code. What if the example code is updated? The developer will need to maintain his example code up-to-date constantly.

This PR is an attempt to fix that and to make example tools modular, so that the developer can later simply link an example lib (lighting-app lib in this case) against his implementation of the two weak methods introduced in this PR.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking given this has been in review for > 3 days.

File: examples/common/datt-config/DattConfig.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <credentials/DeviceAttestationCredsProvider.h>
+#include <credentials/DeviceAttestationVerifier.h>
+#include <credentials/examples/DefaultDeviceAttestationVerifier.h>
+#include <credentials/examples/DeviceAttestationCredsExample.h>
```
Comment (tcarmelveilleux) [inline examples/common/datt-config/DattConfig.h:24]: These includes should not be here, as they cause dependencies to the example

File: examples/platform/linux/AppMain.cpp [hunk 3]
```diff
@@ -25,10 +25,7 @@
 #include <lib/core/CHIPError.h>
 #include <lib/core/NodeId.h>
 
-#include <credentials/DeviceAttestationCredsProvider.h>
-#include <credentials/DeviceAttestationVerifier.h>
-#include <credentials/examples/DefaultDeviceAttestationVerifier.h>
-#include <credentials/examples/DeviceAttestationCredsExample.h>
+#include <DattConfig.h>
```
Comment (tcarmelveilleux) [inline examples/platform/linux/AppMain.cpp:28]: Please use full path

File: examples/platform/linux/BUILD.gn [hunk 4]
```diff
@@ -43,6 +43,8 @@ source_set("app-main") {
     defines += [ "ENABLE_CHIP_SHELL" ]
   }
 
+  deps = [ "${chip_root}/examples/common/datt-config:datt-weak" ]
```
Comment (tcarmelveilleux) [inline examples/platform/linux/BUILD.gn:46]: This makes a platform depend on example code. Examples are not core code

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
The method in this PR causes more usage of global weak linkage, which is not scalable, and does not remove dependencies on examples.

Changes requested:

- [ ] Do not rely on overriding only some globals in examples with weak symbols
- [ ] Split the example DAC provider/verifier into separate build targets in credentials/. It was meant as a starting point, not a forever thing
- [ ] Leave the existing usage of example DAC provider in the examples: examples are not product. You don't go to TEs with examples, you go to TEs with products (which may be based on examples). The examples provide no guarantees of explicit compliance for now.

File: examples/common/datt-config/DattConfig.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+{
+    return chip::Credentials::GetTestAttestationTrustStore();
+}
+
+__attribute__((weak)) chip::Credentials::DeviceAttestationCredentialsProvider * GetAttestationCredentialsProvider()
+{
+    return chip::Credentials::Examples::GetExampleDACProvider();
+}
```
[the remark this answers is not in the document]
Comment (vijs) [reply examples/common/datt-config/DattConfig.cpp:29]: This PR doesn't compromise the example. This PR only adds a new possibility for developers to re-use the current example with their own DAC Provider and TrustedRoot List, by simply adding 2 method definitions that will override those new weak symbols.
In this case, lighting-app will continue to build and act as usual.

[the remark this answers is not in the document]
Comment (cecille) [reply examples/common/datt-config/DattConfig.cpp:29]: I think Tennessee's point, though, is that the proper way to use a platform DAC provider is not to override the example one, but to just set the appropriate credentials provider using the call to SetDeviceAttestationCredentialsProvider, which already exists. The example credentials provider is being used because it is already being explicitly set.

**Comment by vijs** [comment]
Abandoning this PR.
